README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# schtools <a href='http://www.schlosslab.org/schtools/'><img src='man/figures/logo.png' align="right" height="139" /></a>

Schloss Lab Tools for Reproducible Microbiome Research 💩

<!-- badges: start -->

[![check](https://github.com/SchlossLab/schtools/workflows/check/badge.svg)](https://github.com/SchlossLab/schtools/actions)
[![codecov](https://codecov.io/gh/SchlossLab/schtools/branch/master/graph/badge.svg)](https://app.codecov.io/gh/SchlossLab/schtools?branch=main)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/SchlossLab/schtools/blob/main/LICENSE.md)
[![CRAN
status](https://www.r-pkg.org/badges/version/schtools)](https://CRAN.R-project.org/package=schtools)
[![Conda](https://img.shields.io/conda/vn/conda-forge/r-schtools)](https://anaconda.org/conda-forge/r-schtools)
[![DOI](https://zenodo.org/badge/8862218.svg)](https://zenodo.org/badge/latestdoi/8862218)
<!-- badges: end -->

A collection of useful functions and example code created and used by
the Schloss Lab for reproducible microbiome research. Perform common
tasks like read output files from [mothur](https://mothur.org), tidy up
your microbiome data, and format rmarkdown documents for publication.
See the [website](http://www.schlosslab.org/schtools/) for more
information, documentation, and examples.

## Installation

You can install the latest release from
[CRAN](https://cran.r-project.org/package=schtools):
<img src='https://cranlogs.r-pkg.org/badges/grand-total/schtools' align='right'/>

``` r
install.packages("schtools")
```

or the development version from
[GitHub](https://github.com/SchlossLab/schtools) with:

``` r
# install.packages("devtools")
devtools::install_github("SchlossLab/schtools")
```

or install from a terminal using
[conda](https://docs.conda.io/projects/conda/en/latest/index.html) or
[mamba](https://mamba.readthedocs.io/en/latest/):
<img src='https://anaconda.org/conda-forge/r-schtools/badges/downloads.svg' align='right'/>

``` bash
mamba install -c conda-forge r-schtools
```

## Usage

Check out the [introductory
vignette](http://www.schlosslab.org/schtools/articles/introduction.html)
for a quick start tutorial or take a look at the [reference
documentation](http://www.schlosslab.org/schtools/reference/index.html)
for a comprehensive list of functions.

## Help & Contributing

If you have questions, create a new post in
[Discussions](https://github.com/SchlossLab/schtools/discussions).

If you come across a bug, [open an
issue](https://github.com/SchlossLab/schtools/issues) and include a
[minimal reproducible example](https://www.tidyverse.org/help/).

If you’d like to contribute, see our guidelines
[here](http://www.schlosslab.org/schtools/CONTRIBUTING.html).

## Code of Conduct

Please note that the schtools project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## License

The schtools package is licensed under the [MIT
license](https://github.com/SchlossLab/schtools/blob/main/LICENSE.md#mit-license).
The logo is licensed under the [CC BY-SA
license](https://github.com/SchlossLab/schtools/blob/main/LICENSE.md#creative-commons-share-alike-license).

## Citation

To cite schtools in publications, use:

> <p>
> Sovacool KL, Lesniak N, Lucas SK, Armour CR, Schloss PD (2022).
> <em>schtools: Schloss Lab Tools for Reproducible Microbiome
> Research</em>.
> <a href="https://doi.org/10.5281/zenodo.6540686">doi:10.5281/zenodo.6540686</a>,
> <a href="http://www.schlosslab.org/schtools/">http://www.schlosslab.org/schtools/</a>.
> </p>

A BibTeX entry for LaTeX users is:

     @Manual{,
      title = {{schtools}: Schloss Lab Tools for Reproducible Microbiome Research},
      author = {Kelly L. Sovacool and Nicholas Lesniak and Sarah K. Lucas and Courtney R. Armour and Patrick D. Schloss},
      year = {2022},
      doi = {10.5281/zenodo.6540686},
      url = {http://www.schlosslab.org/schtools/},
    } 


NEWS.md
# development version

# schtools 0.4.1

- Minor bug fix in `get_wildcards_tbl()` (#48, @kelly-sovacool).
- Fix roxygen package doc syntax (https://github.com/r-lib/roxygen2/issues/1491, @kelly-sovacool).

# schtools 0.4.0

- `set_knitr_opts()` now sets message to `FALSE` (#45, @kelly-sovacool).
- New function `get_wildcards_tbl()` returns the Snakemake wildcards list as a tibble (#46, @kelly-sovacool).

# schtools 0.3.1

- Fixes for API changes in ggplot2 v3.4.0 and tidyselect v1.2.0 (#42, #44, @kelly-sovacool).
- `pool_taxon_counts()` now takes the `taxon_level` argument as a character 
  rather than a data variable (#43, @kelly-sovacool).

# schtools 0.3.0

- New function `log_snakemake()` redirects all output to a rule-specific log
  file when called by a Snakemake rule (#35, @kelly-sovacool).
- New function `pool_taxon_counts()` pools OTU counts at a particular taxonomic 
  level (e.g. genus, phylum, etc.) to enable comparing analyses at different 
  taxonomic resolutions, as seen in <https://doi.org/10.1128/mbio.03161-21> 
  (#37, @kelly-sovacool).
- New function `calc_relabun()` calculates OTU relative abundances (#38, @kelly-sovacool).
- New custom themes for ggplot2:
    - `theme_sovacool()` (#39, @kelly-sovacool)
    - `theme_lucas()` (#40, @sklucas)

# schtools 0.2.1

- Improvements to `parse_tax()` (#32, @kelly-sovacool).
    - New column `label_html` provides HTML to correctly italicize genus names but not OTU labels.
    - Added an example to the intro vignette to demonstrate using this feature with `ggtext`.
    - Changed taxonomy data frame column names to lowercase and reordered them.
- Minor documentation improvements (#33, @kelly-sovacool).

# schtools 0.1.1

- Refactored `read_tax()` (#30, @kelly-sovacool).
    - New function `parse_tax()` to separate the logic of reading and parsing taxonomy files.
    - `read_tax()` now accepts a `sep` parameter to handle different delim file types (e.g. csv or tsv).
- schtools now has a logo! (#31, @kelly-sovacool)

# schtools 0.1.0

This is the first release version of schtools! 🎉

- You can now install schtools from [CRAN](https://cran.r-project.org/package=schtools)!
- Added a `NEWS.md` file to track changes to the package.
- Major new functions:
  - `read_dist()` (#10, @NLesniak)
  - `read_tax()` (#22, @NLesniak)
  - `set_knitr_opts()` (#21, @pschloss, @kelly-sovacool)
  - `inline_hook()` (#21, @pschloss, @kelly-sovacool)
  - `paste_oxford_list()` (#21, @pschloss, @kelly-sovacool)
  - `format_number()` (#24, @pschloss, @kelly-sovacool)
  - `is_nearly_whole()` (#24, @pschloss, @kelly-sovacool)
  - `close_enough()` (#26, @kelly-sovacool)
  - `is_nondesc()` (#28, @kelly-sovacool)
- New vignettes:
  - `introduction`


R/shared.R
#' Pool OTU counts at a particular taxonomic level
#'
#' Enables comparing analyses at different taxonomic resolutions,
#' as seen in \doi{10.1128/mbio.03161-21}.
#' Implementation adapted from [here](https://github.com/SchlossLab/Armour_Resolution_mBio_2021/blob/master/code/get_phylotype_shared.R).
#'
#' @param otu_shared_dat data frame created from a [shared file](https://mothur.org/wiki/shared_file/)
#'   at the OTU level.
#' @param otu_tax_dat data frame created from a [taxonomy file](https://mothur.org/wiki/taxonomy_file/)
#'  at the OTU level. Must be from the same dataset as the shared file.
#' @param taxon_level taxonomic level to pool OTUs into.
#'   Options: "kingdom", "phylum", "class", "order", "family", "genus".
#'   This should be the name of a column in `otu_tax_dat` as a character string.
#'
#' @return a shared data frame with the OTUs at the specified `taxon_level` and
#'   a corresponding taxonomy dataframe with new OTU numbers.
#' @export
#' @author Kelly Sovacool, \email{sovacool@@umich.edu}
#' @author Pat Schloss \email{pschloss@@umich.edu}
#'
#' @examples
#' tax_dat <- read_tax(system.file("extdata", "test.taxonomy",
#'   package = "schtools"
#' ))
#' shared_dat <- readr::read_tsv(system.file("extdata", "test.shared",
#'   package = "schtools"
#' ))
#' pool_taxon_counts(shared_dat, tax_dat, "genus")
#' pool_taxon_counts(shared_dat, tax_dat, "family")
#' pool_taxon_counts(shared_dat, tax_dat, "phylum")
pool_taxon_counts <- function(otu_shared_dat, otu_tax_dat, taxon_level) {
  countsum <- Group <- label <- numOtus <- otu <- otu_counts <- NULL
  levels <- c("kingdom", "phylum", "class", "order", "family", "genus")
  if (length(taxon_level) != 1) {
    stop(paste0("Only one taxon_level is accepted at a time. You provided: ", length(taxon_level)))
  } else if (!(taxon_level) %in% levels) {
    stop(paste0(
      "taxon_level '", taxon_level,
      "' not recognized. Recognized taxonomic levels are:\n ",
      paste_oxford_list(levels)
    ))
  }
  shared_long <- otu_shared_dat %>%
    tidyr::pivot_longer(tidyr::starts_with("Otu"),
      names_to = "otu",
      values_to = "otu_counts"
    )

  new_otu_nums <- otu_tax_dat %>%
    dplyr::as_tibble() %>%
    dplyr::select(dplyr::all_of(taxon_level)) %>%
    unique() %>%
    dplyr::mutate(otu = paste0(
      "Otu",
      stringr::str_pad(1:nrow(.),
        width = nchar(nrow(.)),
        pad = "0"
      )
    ))

  pooled_shared <- dplyr::inner_join(shared_long,
    otu_tax_dat %>% dplyr::select(otu, dplyr::all_of(taxon_level)),
    by = "otu"
  ) %>%
    dplyr::group_by(!!rlang::sym(taxon_level), Group) %>%
    dplyr::summarize(countsum = sum(otu_counts)) %>%
    dplyr::ungroup() %>%
    dplyr::inner_join(new_otu_nums, by = taxon_level) %>%
    dplyr::select(-dplyr::all_of(taxon_level)) %>%
    tidyr::pivot_wider(names_from = otu, values_from = countsum) %>%
    dplyr::mutate(
      numOtus = ncol(.) - 1,
      label = taxon_level
    ) %>%
    dplyr::select(order(colnames(.))) %>%
    dplyr::select(label, Group, numOtus, tidyr::starts_with("Otu"))

  new_tax <- pooled_shared %>%
    dplyr::select(-c(label, numOtus)) %>%
    tidyr::pivot_longer(tidyr::starts_with("Otu"),
      names_to = "otu",
      values_to = "otu_counts"
    ) %>%
    dplyr::group_by(otu) %>%
    dplyr::summarize(countsum = sum(otu_counts)) %>%
    dplyr::inner_join(new_otu_nums, by = "otu") %>%
    dplyr::rename("size" = "countsum")

  return(list(
    shared = pooled_shared,
    tax = new_tax
  ))
}
